pygsti/algorithms/fiducialselection.py

```diff
@@ -115,20 +147,48 @@ def find_fiducials(target_model, omit_identity=True, eq_thresh=1e-6,
             for gate in fidOps:
                 if frobeniusdist_squared(target_model.operations[gate], Identity) < eq_thresh:
                     fidOps.remove(gate)
-
-        availableFidList = _circuits.list_all_circuits(fidOps, 0, max_fid_length)
         
-        circuit_cache= create_circuit_cache(target_model,availableFidList)
+        availableFidList = []
+        if max_fid_length is not None:
+            if candidate_fid_counts is not None:
+                warn('Specifying max_fid_length without setting candidate_fid_counts results the settings specified in candidate_fid_counts being overridden.\
+                candidate_fid_counts is set by defaults, so this might be expected, but if this is not the desired behavior then please verify your arguments.')
+            #now overrive the value of candidate_fid_counts
+            candidate_fid_counts= max_fid_length
+        if isinstance(candidate_fid_counts, int): 
+            candidate_fid_counts = {candidate_fid_counts: 'all upto'}
+        
+        for fidLength, count in candidate_fid_counts.items():
+            if count == "all upto":
+                availableFidList.extend(_circuits.list_all_circuits(fidOps, 0, fidLength))
+            #if "all" then include all circuits at that particular length.
+            elif count =="all":
+                availableFidList.extend(_circuits.list_all_circuits_one_len(fidOps, fidLength))
+            else:
+                availableFidList.extend(_circuits.list_random_circuits_onelen(
+                    fidOps, fidLength, count, seed=candidate_seed))
         
         printer.log('Initial Length Available Fiducial List: '+ str(len(availableFidList)), 1)
+        
+        
+        printer.log('Creating cache of fiducial process matrices.', 3)
+        circuit_cache= create_circuit_cache(target_model,availableFidList)
+        printer.log('Completed cache of fiducial process matrices.', 3)
+        
         #print('Initial Length Available Fiducial List: ', len(availableFidList))
         
         #Now that we have a cache of PTMs as numpy arrays for the initial list of available fiducials
         #we can clean this list up to remove any effective identities and circuits with duplicate effects.
-        
-        cleaned_availableFidList, cleaned_circuit_cache = clean_fid_list(target_model, circuit_cache, availableFidList,
-                                                                        drop_identities=True, drop_duplicates=True,
-                                                                        eq_thresh=eq_thresh)
+        #Use a flag to check if we can assume these are clifford circuits in which case
+        #we can use a more efficient deduping routine.
+        if assume_clifford:
```

Comment:
Could be cleaned up IMO - see below comment in clean_fid_list_clifford.

Reply:
Sounds good, I refactored this in line with your suggestions.